### PR TITLE
Kubevirt List and Get VmiPreset

### LINKED ISSUE #1114: Add API to get/list Virtual Machine Presets for dashboard

### PR DESCRIPTION
**Pull request by hdurand0710**
Signed-off-by: Helene Durand <helene@kubermatic.com>

**What this PR does / why we need it**:

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes kubermatic/machine-controller#1114

**Special notes for your reviewer**:
This PR adds the API endpoints to list:
- Kubevirt VirtualMachineInstancePreset
- StorageClass
They have been added in v2, with credentials provided in the request or credentials from the cluster.

**Documentation**:

**Does this PR introduce a user-facing change?**:

```release-note
Add API endpoints for Kubevirt VirtualMachineInstancePresets and StorageClasses
```

### CODE REVIEW

**Comment by mfranczy** [comment]
blocked by https://github.com/kubevirt/kubevirt/pull/6938

**Comment by mfranczy** [comment]
related https://github.com/kubermatic/machine-controller/issues/1112

**Comment by hdurand0710** [comment]
related https://github.com/kubermatic/machine-controller/issues/1114 
(related https://github.com/kubermatic/machine-controller/issues/1112 was implemented by https://github.com/kubermatic/kubermatic/pull/8360
)

File: hack/verify-spelling.sh
Comment (mfranczy) [inline hack/verify-spelling.sh]: please don't forget to remove it once the upstream PR is merged

File: pkg/api/v2/types.go
Comment (mfranczy) [inline pkg/api/v2/types.go]: cannot we just make an alias of KubeVirt type (which is not a pointer)? Instead of creating a new one with embedded struct?
```
type  VirtualMachineInstancePresetList kubevirtv1.VirtualMachineInstancePresetList
```

Comment (mfranczy) [reply pkg/api/v2/types.go]: that applies to below structs too

File: pkg/api/v2/types.go
Comment (mfranczy) [inline pkg/api/v2/types.go]: I know I suggested the first version in the past :) but...
let's change:
`VirtualMachineInstancePreset represents a preset of virtual machine instance`
to 
`VirtualMachineInstancePreset represents a KubeVirt Virtual Machine Instance Preset`

File: pkg/api/v2/types.go
Comment (mfranczy) [inline pkg/api/v2/types.go]: let's change to 
> // StorageClassList represents a list of Kubernetes StorageClass.

File: pkg/handler/common/provider/kubevirt.go
Comment (mfranczy) [inline pkg/handler/common/provider/kubevirt.go]: 2021 :)

**Review by mfranczy — COMMENTED** [review]
I left some minor comments, I will do a second review round once you adjust tests as we discussed, Thanks @hdurand0710

File: pkg/api/v2/types.go
Comment (hdurand0710) [reply pkg/api/v2/types.go]: Nope, unfortunately, I tried to directly make an alias. It makes the swagger client generation crashes:
% ./hack/gen-api-client.sh
[2021-12-13T15:11:18+01:00] Running ./hack/gen-api-client.sh in a Docker container using golang:1.17.1...
2021/12/13 14:12:20 WARNING: can't figure out object type for embedded named type (*types.Named): string
2021/12/13 14:12:20 WARNING: can't figure out object type for embedded named type (*types.Named): string
signal: killed

It's reproductible. With the struct and pointer, it works.

File: pkg/api/v2/types.go
Comment (hdurand0710) [reply pkg/api/v2/types.go]: Done

File: pkg/api/v2/types.go
Comment (hdurand0710) [reply pkg/api/v2/types.go]: done

File: hack/verify-spelling.sh
Comment (hdurand0710) [reply hack/verify-spelling.sh]: Done already. The script will fail until the upstream PR is merged. Won't forget then 👍

**Comment by hdurand0710** [comment]
@mfranczy pushed with the first round of comments integrated.

**Comment by hdurand0710** [comment]
@mfranczy : updated with upstream, spellcheck is now ok.
You can have a look when you have time.

**Comment by hdurand0710** [comment]
@zreigz I've updated according to our discussions, if you can review. thanks!

File: pkg/handler/common/provider/kubevirt.go
Comment (zreigz) [inline pkg/handler/common/provider/kubevirt.go]: you can use
```
import metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"

metav1.NamespaceDefault

```

File: pkg/handler/common/provider/kubevirt.go
Comment (zreigz) [inline pkg/handler/common/provider/kubevirt.go]: what if you get a malformed kubeconfig? In your case, you can't recognize it. You can check if the kubeconfig is base64 and then decode and throw the error if happened
```
func isBase64(input string) bool {
	r, _ := regexp.Compile(`^([A-Za-z0-9+/]{4})*([A-Za-z0-9+/]{3}=|[A-Za-z0-9+/]{2}==)?$`)
	return r.Match([]byte(line))
}
```

File: pkg/handler/v2/provider/kubevirt.go
Comment (zreigz) [inline pkg/handler/v2/provider/kubevirt.go]: some leftover ?

File: pkg/handler/v2/provider/kubevirt_test.go
Comment (zreigz) [inline pkg/handler/v2/provider/kubevirt_test.go]: It's really hard to read the test scenarios. Can you split the tests for different endpoints like for other providers:
 - api/v2/providers/kubevirt/vmipresets
 - /api/v2/projects/%s/clusters/%s/providers/kubevirt/vmipresets

**Review by zreigz — CHANGES_REQUESTED** [review]

File: pkg/handler/common/provider/kubevirt.go
Comment (zreigz) [reply pkg/handler/common/provider/kubevirt.go]: Second thought. You send kubeconfig in the header in the endpoint. I think it always must be base64. Please verify this. We also use base64 for sending kubeconfig in other endpoints, if true then `config = []byte(kubeconfig)` it's not necessary

File: pkg/handler/common/provider/kubevirt.go
Comment (hdurand0710) [reply pkg/handler/common/provider/kubevirt.go]: Thanks, done.

File: pkg/handler/common/provider/kubevirt.go
Comment (hdurand0710) [reply pkg/handler/common/provider/kubevirt.go]: It's true that it will always be base64 encoded. I have updated accordingly.

File: pkg/handler/v2/provider/kubevirt.go
Comment (hdurand0710) [reply pkg/handler/v2/provider/kubevirt.go]: :-) right, removed.

File: pkg/handler/v2/provider/kubevirt_test.go
Comment (hdurand0710) [reply pkg/handler/v2/provider/kubevirt_test.go]: I've re-pushed with the different endpoints splitted.

**Comment by hdurand0710** [comment]
@zreigz , I have integrated your comments. I let you have a look when you have time.

File: pkg/api/v2/types.go
Comment (zreigz) [inline pkg/api/v2/types.go]: you can remove `protobuf` annotations

File: pkg/handler/v2/routes_v2.go
Comment (zreigz) [inline pkg/handler/v2/routes_v2.go]: some unexpected change?

Comment (hdurand0710) [reply pkg/handler/v2/routes_v2.go]: oh wow!!!!! so unexpected yes ! No idea where this comes from. wow!!!

File: pkg/api/v2/types.go
Comment (hdurand0710) [reply pkg/api/v2/types.go]: done

**Comment by hdurand0710** [comment]
@zreigz , ready for next round of review. Thanks.

**Comment by hdurand0710** [comment]
@zreigz : can you please have another look? @maciaszczykm has done some changes:
- we keep only the `list` and removed the `get`
- naming conventions
Thanks

**Comment by mfranczy** [comment]
> @zreigz : can you please have another look? @maciaszczykm has done some changes:

actually @mfranczy :)

@hdurand0710 last thing, please change the release note to something shorter like:
> Add API endpoints for Kubevirt VirtualMachineInstancePresets and StorageClasses

**Comment by hdurand0710** [comment]
> > @zreigz : can you please have another look? @maciaszczykm has done some changes:
> 
> actually @mfranczy :)

Ooupsi !
> 
> @hdurand0710 last thing, please change the release note to something shorter like:
> 
> > Add API endpoints for Kubevirt VirtualMachineInstancePresets and StorageClasses

Done

**Review by mfranczy — APPROVED** [review]
/approve
/lgtm

**Pull request merged**
